pattern.c

```diff
@@ -237,6 +238,80 @@ static bool eat_regex(struct Pattern *pat, struct Buffer *s, struct Buffer *err)
   return true;
 }
 
+/**
+ * eat_query - Parse a query for an external search program
+ * @param pat  Pattern to match
+ * @param s   String to parse
+ * @param err Buffer for error messages
+ * @retval true If the pattern was read succesfully
+ */
+static bool eat_query(struct Pattern *pat, struct Buffer *s, struct Buffer *err)
+{
+  struct Buffer cmd_buf;
+  struct Buffer tok_buf;
+  int result;
+  FILE *fp = NULL;
+  pid_t pid;
+  char* oldline = NULL;
+  char* newline;
+  char* dupline;
+  char* nows;
+  size_t lsz;
+
+  if (!(SearchCommand && *SearchCommand)) {
+    mutt_buffer_printf(err, "%s", _("No search command defined"));
+    return false;
+  }
+
+  char *pexpr = s->dptr;
+  mutt_buffer_init(&tok_buf);
+  result = mutt_extract_token(&tok_buf, s, MUTT_TOKEN_PATTERN | MUTT_TOKEN_COMMENT);
+  if (result != 0 || !tok_buf.data)
+  {
+    mutt_buffer_printf(err, _("Error in expression: %s"), pexpr);
+    return false;
+  }
+  if (!*tok_buf.data)
+  {
+    mutt_buffer_printf(err, "%s", _("Empty expression"));
+    FREE(&tok_buf.data);
+    return false;
+  }
+
+  mutt_buffer_init(&cmd_buf);
+  mutt_buffer_addstr(&cmd_buf, SearchCommand);
+  mutt_buffer_addch(&cmd_buf, ' ');
+  mutt_buffer_addstr(&cmd_buf, tok_buf.data);
+  FREE(&tok_buf.data);
+
+  mutt_message(_("Running search command..."));
+  pat->ismulti = 1;
+  mutt_list_clear(&pat->p.multi_cases);
+  pid = mutt_create_filter(cmd_buf.data, NULL, &fp, NULL);
+  if (pid < 0)
+  {
+    mutt_buffer_printf(err, "unable to fork command: %s\n", cmd_buf.data);
+    FREE(&cmd_buf.data);
+    return false;
+  }
+  
+  while ((newline = mutt_file_read_line(oldline, &lsz, fp, NULL, 0)) != NULL) {
```

Comment:
For sake of readability, i think it would be best to move the assignment out of the conditional. It'd result in duplication, but will require less cognitive load to interpret the program's behavior.

Reply:
Let me just say that I see the tradeoff and I choose the other evil LOL

Maybe if doing it your way would result in other improvements (like saving a variable or 2), I might be swayed.  So I will think about it.

Reply:
I'm only looking out for myself here :P I __know__ i will skim over the assignment-in-conditional and be confused at some point in the future :)